tools/rpc.toit

```diff
@@ -13,133 +13,141 @@
-  // The third argument must be the name.
-  register_named_procedure name/string procedure_name/int handler/Lambda:
-    map := named_handlers_.get procedure_name --init=:{:}
-    map[name] = handler
+  register_procedure name/int action/Lambda -> none:
+    procedures_[name] = action
 
   // Register a descriptor-based procedure to handle a message.  These are
   // invoked by the RPC caller with a descriptor as the first argument.  This
   // descriptor is looked up on the process group and the resulting object is
   // passed to the handler.
-  register_descriptor_procedure procedure_name/int handler/Lambda:
-    descriptor_handlers_[procedure_name] = handler
+  register_descriptor_procedure name/int action/Lambda:
+    handlers_[name] = action
 
-  get_descriptor_ gid context:
+  // Typically overwritten in a subclass.
+  get_descriptor_ gid/int descriptor/int -> any:
     return null
 
-  get_handler_ gid procedure_name procedure_args [--on_error]:
-    handlers_.get procedure_name --if_present=: return it
-
-    if procedure_args.size == 0:
-      on_error.call "Missing call context"
-      return null
-
-    context ::= procedure_args[0]
-
-    named_handlers_.get procedure_name --if_present=:
-      it.get context --if_present=: return it
-
-    // The descriptor must be an integer.
-    if context is not int:
-      on_error.call "Closed descriptor $context"
-      return null
-
-    descriptor_handlers_.get procedure_name --if_present=: | handler |
-      driver := get_descriptor_ gid context
-      if driver: return :: | args gid pid |
-        handler.call driver args gid
-
-      on_error.call "Closed descriptor $context"
-      return null
-
-    on_error.call "No such procedure registered $procedure_name"
-    return null
-
-  on_message type gid pid args:
-    if type == SYSTEM_RPC_CHANNEL_LEGACY_:
-      on_open_channel_ gid pid args
-      return
-
-  static is_rpc_error_ error -> bool:
-    if error == Channel_.CHANNEL_CLOSED_ERROR: return true
-    if error == Channel_.NO_SUCH_CHANNEL_ERROR: return true
-    return false
-
-  on_open_channel_ gid pid id_bytes:
-    task::
-      catch --trace=(: not is_rpc_error_ it):
-        channel := Channel_.open (Uuid id_bytes)
+class RpcRequest_:
+  next/RpcRequest_? := null
+
+  pid/int
+  gid/int
+  id/int
+  arguments/List
+  procedure/Lambda
+
+  constructor .pid .gid .id .arguments .procedure:
+
+  process -> none:
+    result/any := null
+    try:
+      result = procedure.call arguments gid pid
+      if result is Serializable: result = result.serialize
+    finally: | is_exception exception |
+      reply := is_exception
+          ? [ id, true, exception.value, exception.trace ]
+          : [ id, false, result ]
+      process_send_bytes_ pid SYSTEM_RPC_CHANNEL_LEGACY_ (ubjson.encode reply)
+      return  // Stops any unwinding.
+
+monitor RpcRequestQueue_:
+  static MAX_TASKS ::= 4
+  static MAX_REQUESTS ::= 16
+  static IDLE_TIME_MS ::= 10_000
```

Comment:
I think this is too long.  Compare with the time to do a GC caused by the new task being created, which is a few ms.  We want the extra tasks to be gone by the time the next memory high water arrives.